Vision/PID/diagramskill/app/__init__.py

```diff
@@ -0,0 +1,491 @@
+            start_time = time.time()
+            boxes,matches,singles = match_boxes(textblocks)
+            boxes = match_singles(boxes,matches,singles)
+            logging.info("Group Text--- %s seconds ---" % (time.time() - start_time))
+
+            text_array.append(get_text(boxes))
+
+        start_time = time.time()
+        text = ' '.join(text_array)    
+
+        text = text.replace("- ","-") 
+        text = text.replace(" -","-") 
+        strings = text.split(" ")
+        if len(strings) >0:
+            block_array = [ x for x in strings if "-" not in x]
+            if len(block_array) > 0: block = ' '.join(block_array)
+            tags_array = [ x for x in strings if "-"  in x and len(x)>1]
+            if len(tags_array) > 0: tags = ' '.join(tags_array)
+        else:
+            block = ""
+            tags = ""  
+
+        logging.info("Finalize Text--- %s seconds ---" % (time.time() - start_time))
+
+
+    except Exception as ex:
+        template = "An exception of type {0} occurred. Arguments:\n{1!r}"
+        message = template.format(type(ex).__name__, ex.args)
+        logging.error(message)
+
+
+    results['values'].append({"recordId": recordId,"data": { "textblocks": block,"tags":tags} })
+    
+    
+
+ 
+def get_stream(img,format='JPEG'):
+
+    """
+    Convert CV2 Image to ByteIO Stream
+
+    Parameters::
+      (CV2 Image): img -- OpenCV Image
+      (String) : format -- image format (Default JPEG)
+
+    Returns:
+       ByteIO Steam
+    """
+    logging.info("calling get_stream")
+    pil_image2 = Image.fromarray(cv2.cvtColor(img,cv2.COLOR_BGR2RGB))
+    buf = BytesIO()
+    pil_image2.save(buf, format=format)
+    buf.seek(0)
+    return buf
+
+
+def get_text(boxes):
+    """
+    Merge text from boxes (Textblock)
+
+    Parameters::
+      (TextBlock[]): boxes -- OCR boundingbox grouping to retreive test from 
+
+    Returns:
+        TextS (String): concat textblock text
+    """
+    text = ""
+    for b in boxes:
+        text = text + " " +  b.text 
+    return text
+
+
+
+def match_singles(boxes,matches,singles):
+    """
+    Convert CV2 Image to ByteIO Stream
+
+    Parameters::
+      (CV2 Image): img -- OpenCV Image
+      (String) : format -- image format (Default JPEG)
+
+    Returns:
+       ByteIO Steam
+    """
+    for s in singles:
+        cnt = 0
+        for m in matches:
+            if(matches[m].intersect(singles[s]) == True):
+                cnt = 0
+                break
+            else:
+                cnt = 1
+        
+        if cnt >0:
+         boxes.append(singles[s])
+
+    return boxes
+
+
+
+def match_boxes(textblocks):
+    boxes = []
+    matches = {}
+    ids=[]
+    singles = {}
+    maxSegment = 95
+    leftAlignSensitivity = 5
+    rightAlignSensitivity = 5
+    centerAlignSensitivty = 15
+ 
+    for i in range(len(textblocks)):
+        tb = textblocks[i].copy()
+        count=0
+        for j in range(len(textblocks)):
+            if i != j:

```

Comment:
Those nested ifs look like they should be AND'ed

Reply:
works as is... :)

Reply:
I'm sure it does, it's just that it would be more readable.